linux/src/threadpool_linux.c

```diff
@@ -0,0 +1,516 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+
+#include <stdlib.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <semaphore.h>
+#include <time.h>
+#include <string.h>
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/xlogging.h"
+
+#include "c_pal/gballoc_hl.h" // IWYU pragma: keep
+#include "c_pal/gballoc_hl_redirect.h"
+#include "c_pal/threadapi.h"
+#include "c_pal/interlocked.h"
+#include "c_pal/sync.h"
+#include "c_pal/srw_lock.h"
+#include "c_pal/execution_engine.h"
+#include "c_pal/execution_engine_linux.h"
+
+#include "c_pal/threadpool.h"
+
+#define TIMEOUT_MS          100
+#define MAX_THREAD_COUNT    1
+#define THREAD_LIMIT        20
+
+#define TASK_RESULT_VALUES  \
+    TASK_NOT_USED,          \
+    TASK_INITIALIZING,      \
+    TASK_WAITING,           \
+    TASK_WORKING
+
+MU_DEFINE_ENUM_WITHOUT_INVALID(TASK_RESULT, TASK_RESULT_VALUES);
+MU_DEFINE_ENUM_STRINGS(TASK_RESULT, TASK_RESULT_VALUES)
+
+#define THREADPOOL_STATE_VALUES \
+    THREADPOOL_STATE_NOT_OPEN,  \
+    THREADPOOL_STATE_OPENING,   \
+    THREADPOOL_STATE_OPEN,      \
+    THREADPOOL_STATE_CLOSING
+
+MU_DEFINE_ENUM(THREADPOOL_STATE, THREADPOOL_STATE_VALUES)
+
+MU_DEFINE_ENUM_STRINGS(THREADPOOL_STATE, THREADPOOL_STATE_VALUES)
+
+typedef struct THREADPOOL_TASK_TAG
+{
+    volatile_atomic int32_t task_state;
+    THREADPOOL_WORK_FUNCTION task_func;
+    void* task_param;
+} THREADPOOL_TASK;
+
+typedef struct THREADPOOL_TAG
+{
+    volatile_atomic int32_t state;
+    volatile_atomic int32_t thread_count;
+    uint32_t max_thread_count;
+    uint32_t min_thread_count;
+    int32_t used_thread_count;
+    volatile_atomic int32_t quitting;
+    volatile_atomic int32_t task_count;
+
+    sem_t semaphore;
+
+    // Pool Array
+    THREADPOOL_TASK* task_array;
+    volatile_atomic int64_t insert_idx;
+    volatile_atomic int64_t consume_idx;
+    volatile_atomic int32_t task_array_size;
+    volatile_atomic int32_t pending_call_count;
+
+    SRW_LOCK_HANDLE srw_lock;
+    uint32_t list_index;
+
+    THREAD_HANDLE* thread_handle_array;
+} THREADPOOL;
+
+static void internal_close(THREADPOOL_HANDLE threadpool)
+{
+    bool should_wait_for_transition = false;
+    do
+    {
+        THREADPOOL_STATE current_state = interlocked_compare_exchange(&threadpool->state, THREADPOOL_STATE_CLOSING, THREADPOOL_STATE_OPEN);
+        if (current_state == THREADPOOL_STATE_NOT_OPEN || current_state == THREADPOOL_STATE_CLOSING)
+        {
+            should_wait_for_transition = false;
+        }
+        else if (current_state == THREADPOOL_STATE_OPENING)
+        {
+            should_wait_for_transition = true;
+        }
+        else
+        {
+            (void)interlocked_exchange(&threadpool->quitting, 1);
+
+            while (interlocked_add(&threadpool->pending_call_count, 0) != 0)
+            {
+                ThreadAPI_Sleep(1);
```

Comment:
>ThreadAPI_Sleep(1) [](http://example.com/codeflow?start=16&length=18)

wait_on_address?

Reply:
The wait_on_address in linux seems inconsistent

I have a screenshot of the call using wait_on_address with as pending_call_count already == 0 and then it sitting on the wait_on_address.  The code looks like this:
// threadpool->pending_call_count is already == zero here
            do
            {
                (void)wait_on_address(&threadpool->pending_call_count, 0, UINT32_MAX);  // The code just sits in here
            } while (&threadpool->pending_call_count != 0);

I believe we'll need to look at this code for linux, I don't know why this happens and it will require some research into what this call is doing.  This is why I did this interlock_add here.

Reply:
That's a problem because we use wait_on_address in other places so it sounds like more will be broken on Linux, we need to fix that (separate task).